### ISSUE TITLE
Do something more reasonable when 00-index.tar is non-security but you try to cabal update with security enabled

### ISSUE THREAD

**Issue by ezyang** [haskell/cabal#3854]
Steps to reproduce:

1. You'll need two ways to invoke Cabal, one where Hackage security is on and another where it is off. Use a HOME directory with no .cabal
2. Run cabal update with the security turned off. This will give you 00-index.tar with no security
3. Run `cabal update -v3` with security turned on

You'll notice that in the invocation of curl has a `Range` header, even though it shouldn't, because while it's true we have some existing tarball, we absolutely cannot reuse it.

@hvr says that this is just a good reason for us to NOT put the secure tarball in 00-index.tar, clobbering the old insecure tarball. See  https://github.com/well-typed/hackage-security/issues/117

CC @dcoutts @edsko

**Comment by edsko** [comment]
You haven't shown the full behaviour, so I don't know if this is actually a bug or not. Here is what is supposed to happen:

1. It attempts to incrementally update the 00-index
2. The hash gets verified; in your case, this will of course fail
3. It retries; it updates the root information, and attempts an incremental update again
4. The hash gets verified; this will fail once more. At this point, update will fail with a _different_ error:

``` haskell
    -- | Update failed twice
    --
    -- If we attempt an incremental update the first time, and it fails,  we let
    -- it go round the loop, update local security information, and try again.
    -- But if an incremental update then fails _again_, we  instead attempt a
    -- regular download.
  | UpdateFailedTwice
```

At this point, we will do a non-incremental update of the index, and all will be well. 

In other words, this is only a bug if we keep retrying forever, which actually should never happen at all. 

PS. Can we please avoid inventing new "hells" for people to complain about?

**Comment by ezyang** [comment]
I'm currently on some "less good" wifi, and so after waiting half an hour for the index to download (and curl timing out twice!), I gave up, deleted all of my old package metadata, and did a clean download (which did finish in about 15 minutes).

**Comment by hvr** [comment]
I'll try improving the situation (this has ramifications for the `--index-state` feature and therefore it's desirable for me to have cabal-install know whether it's reading a 01-index or a 00-index format)
